File: solutions/problem_1071_greatest_common_divisor_of_strings/solution.py
"""
Problem 1071: Greatest Common Divisor of Strings
==============================================
Difficulty: Easy
Tags: Math, String
Pattern: Math / String

Time Complexity:  O(n+m)
Space Complexity: O(n+m)
"""

from math import gcd

class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if str1 + str2 != str2 + str1:
            return ''
        return str1[:gcd(len(str1), len(str2))]

File: solutions/problem_1071_greatest_common_divisor_of_strings/solution.cpp
/*
 * Problem 1071: Greatest Common Divisor of Strings
 * ==============================================
 * Difficulty: Easy
 * Tags: Math, String
 * Pattern: Math / String
 *
 * Time Complexity:  O(n+m)
 * Space Complexity: O(n+m)
 */

#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string gcdOfStrings(string& str1, string& str2) {
        // Mathematical approach
        long long result = 0;
        int x = str1;
        while (x != 0) {
            result = result * 10 + x % 10;
            x /= 10;
        }
        return (int)result;
    }
};
